### PR TITLE
Support Composer 2

### PR DESCRIPTION
**Pull request by mglaman**
```
  Problem 1
    - civicrm/civicrm-core[5.29.0, ..., 5.31.x-dev] require civicrm/composer-downloads-plugin ^2.0 -> satisfiable by civicrm/composer-downloads-plugin[v2.0.0, v2.1.0, v2.1.1].
    - civicrm/composer-downloads-plugin[v2.0.0, ..., v2.1.1] require composer-plugin-api ^1.1 -> found composer-plugin-api[2.0.0] but it does not match the constraint.
    - Root composer.json requires civicrm/civicrm-core ~5.29 -> satisfiable by civicrm/civicrm-core[5.29.0, ..., 5.31.x-dev].
```

### CODE REVIEW

**Comment by mglaman** [comment]
This needs manual testing, I thought there was CI attached.

**Comment by seamuslee001** [comment]
@mglaman My unerstanding from @totten is that there needs to be some re-writing of this plugin because the underlying API within composer has changed from v1 => v2

**Comment by mglaman** [comment]
Thanks, @seamuslee001, I assumed there would be _something_ which broke. I'll see if I can work on the changes and support v1 and v2 of the API

**Comment by mglaman** [comment]
Okay, I've added an Actions workflow and running the tests locally.

I'll work on fixing `PharHandler` where it's getting an incorrect directory for `chmod`

**Comment by mglaman** [comment]
Looks like it should be passing, now: https://github.com/mglaman/composer-downloads-plugin/actions/runs/336596453

File: src/Handler/ArchiveHandler.php [hunk 1]
```diff
@@ -2,9 +2,11 @@
 
 namespace LastCall\DownloadsPlugin\Handler;
 
+use Clue\React\Block;
```
Comment (mglaman) [inline src/Handler/ArchiveHandler.php:5]: Argh this is an unused import from a library I used to wait for a promise to resolve until I realized Composer had a helper method.

File: src/Handler/ArchiveHandler.php [hunk 2]
```diff
@@ -2,9 +2,11 @@
...
 use Composer\Composer;
 use Composer\IO\IOInterface;
 use LastCall\DownloadsPlugin\GlobCleaner;
+use React\Promise\PromiseInterface;
```
Comment (mglaman) [inline src/Handler/ArchiveHandler.php:9]: Unused import

File: src/Handler/FileHandler.php [hunk 3]
```diff
@@ -41,22 +42,34 @@ public function download(Composer $composer, IOInterface $io) {
...
         $pkg = clone $this->getSubpackage();
         $pkg->setTargetDir($tmpDir);
         $downloadManager = $composer->getDownloadManager();
-        $downloadManager->download($pkg, $tmpDir);
+        $promise = $downloadManager->download($pkg, $tmpDir);
 
-        if (file_exists($target)) {
-            $cfs->remove($target);
+        // composer:v2
+        if ($promise instanceof PromiseInterface) {
```
Comment (mglaman) [inline src/Handler/FileHandler.php:55]: This should use version_compare like the archive downloader.

File: tests/SniffTest.php [hunk 4]
```diff
@@ -53,7 +53,8 @@ public static function getComposerJson() {
     public static function setUpBeforeClass() {
         parent::setUpBeforeClass();
         self::initTestProject(static::getComposerJson());
-        PH::runOk('composer install -v');
+      $composer_path = self::getComposerPath();
```
Comment (mglaman) [inline tests/SniffTest.php:56]: Indentation needs fixing

**Comment by seamuslee001** [comment]
I've tested this on both a composer2 and composer1 install and have run unit tests with both and confirmed it works, I'm going to merge this and tag a new release

**Comment by mglaman** [comment]
@seamuslee001 thanks! I'll try it out on my CI runs which install civicrm-core

**Pull request merged**